Add Nystromformer blog post (#414)
huggingface/blog: nystromformer.md

---
title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"

> thumbnail: /blog/assets/86_nystromformer/paper_figure.png

Reviewer: Totally! Figma is a good choice, or you could use Canva instead.

Here's a ZIP file with a bunch of SVGs that you could choose from for a background, and a HF logo: https://drive.google.com/file/d/1d-uLYKWUV1K10MRmnxXfW6MAUrTd4qaT/view?usp=sharing

And the font we use is "Source Sans Pro", bold.

> thumbnail: /blog/assets/86_nystromformer/paper_figure.png

Author: Thanks for sharing this. I've pushed the thumbnail I made to this branch. Please let me know what you think.

---
title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
thumbnail: /blog/assets/86_nystromformer/paper_figure.png
---

<h1>Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method</h1>

<script async defer src="https://unpkg.com/medium-zoom-element@0/dist/medium-zoom-element.min.js"></script>

## Introduction

Transformers have exhibited remarkable performance on various Natural Language Processing and Computer Vision tasks. Their success can be attributed to the self-attention mechanism, which captures the pairwise interactions between all the tokens in an input. However, the standard self-attention mechanism has a time and memory complexity of \\(O(n^2)\\) (where \\(n\\) is the length of the input sequence), making it expensive to train on long input sequences. 

> The [Nyströmformer](https://arxiv.org/abs/2102.03902) is one of many efficient transformer models that approximates standard self-attention with \\(O(n)\\) complexity. Nyströmformer exhibits competitive performance on various downstream NLP tasks while improving upon the efficiency of standard self-attention. The aim of this blog post is to give readers an overview of the Nyström method and how it can be adapted to approximate self-attention.

Reviewer 2: ```suggestion
The [Nyströmformer](https://arxiv.org/abs/2102.03902) is one of many efficient Transformer models that approximates standard self-attention with \\(O(n)\\) complexity. Nyströmformer exhibits competitive performance on various downstream NLP tasks while improving upon the efficiency of standard self-attention. The aim of this blog post is to give readers an overview of the Nyström method and how it can be adapted to approximate self-attention.
```
I always capitalize Transformer :D btw the Nyströmformer author also showed that it works well on ImageNet, right?

Author: Yeah, it looks better capitalized. I shall make that change everywhere. I don't think the author ran ImageNet experiments, but there is an image classification task in the LRA benchmark.

Reviewer 2: Yeah I was referring to this: https://github.com/mlpen/Nystromformer/tree/main/ImageNet.

But doesn't seem like he reported any metrics

Reviewer 2: Update: from the README:
> We extended segment-means to compute landmarks without requiring the sequence length divisible by the number of landmarks. Then we used this Nystromformer to perform deployment of T2T-Vit_t-14 for image classification without retraining. Our T2T-ViT-Nys-14 achieves 78% top-1 accuracy, outperforming performer/Linformer +4.3%/+12.7% for the direct deployment.

Author: Oh yeah, I forgot about that. I changed 

> various downstream NLP tasks

to

> various downstream NLP and CV tasks

using segment means. In this procedure, \\(n\\) tokens are grouped into \\(m\\) segments, and the mean of each segment is computed. Ideally, \\(m\\) is much smaller than \\(n\\). According to experiments from the paper, selecting just \\(32\\) or \\(64\\) landmarks produces competetive performance compared to standard self-attention and other efficient attention mechanisms, even for long sequences lengths (\\(n = 4096\\) or \\(8192\\)). 

The overall algorithm is summarised by the following figure from the paper:

<figure class="image table text-center m-0 w-full">
  <medium-zoom background="rgba(0,0,0,.7)" alt="Efficient self-attention with the Nyström method" src="assets/86_nystromformer/paper_figure.png"></medium-zoom>
  <figcaption>Efficient self-attention with the Nyström method</figcaption>
</figure>

The three orange matrices above correspond to the three matrices we constructed using the key and query landmarks. Also, notice that there is a DConv box. This corresponds to a skip connection added to the values using a 1D depthwise convolution.

## How is Nyströmformer implemented?

> The original implementation of Nyströmformer can be found [here](https://github.com/mlpen/Nystromformer) and the HuggingFace implementation can be found [here](https://github.com/huggingface/transformers/tree/main/src/transformers/models/nystromformer). Let's take a look at a few lines of code (with some comments added) from the HuggingFace implementation. Note that some details such as normalization, attention masking, and depthwise convolution are avoided for simplicity.

Reviewer 2: ```suggestion
The original implementation of Nyströmformer can be found [here](https://github.com/mlpen/Nystromformer) and the HuggingFace implementation can be found [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/nystromformer/modeling_nystromformer.py). Let's take a look at a few lines of code (with some comments added) from the HuggingFace implementation. Note that some details such as normalization, attention masking, and depthwise convolution are avoided for simplicity.
```

```python
key_layer = self.transpose_for_scores(self.key(hidden_states)) # K
value_layer = self.transpose_for_scores(self.value(hidden_states)) # V
query_layer = self.transpose_for_scores(mixed_query_layer) # Q

.
.

> .

Reviewer 2: What's the purpose of these 3 dots? :D

Author: I thought they would signify that I left out some part of the code. But, on second thought, they seem unnecessary. I shall remove them.

```python
from transformers import AutoTokenizer, NystromformerForMaskedLM
import torch

tokenizer = AutoTokenizer.from_pretrained("uw-madison/nystromformer-512")
model = NystromformerForMaskedLM.from_pretrained("uw-madison/nystromformer-512")

inputs = tokenizer("Paris is the [MASK] of France.", return_tensors="pt")

with torch.no_grad():
    logits = model(**inputs).logits

# retrieve index of [MASK]
mask_token_index = (inputs.input_ids == tokenizer.mask_token_id)[0].nonzero(as_tuple=True)[0]

predicted_token_id = logits[0, mask_token_index].argmax(axis=-1)
tokenizer.decode(predicted_token_id)

> ```

Reviewer 2: Would be nice to include the output here as well, you can check other blog posts to see how the output is added, e.g. https://huggingface.co/blog/how-to-generate

Author: Thanks for the suggestion. Added output for both code blocks.

> Alternatively, we can use pipeline:

Reviewer 2: ```suggestion
Alternatively, we can use the pipeline API (which handles all the complexity for us):
```

---
title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
thumbnail: /blog/assets/86_nystromformer/thumbnail.png
---

> 

Reviewer 3: It would be great if you could add a data + an author card as in https://github.com/huggingface/blog/blob/main/1b-sentence-embeddings.md?plain=1#L9-L25

---
title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
thumbnail: /blog/assets/86_nystromformer/thumbnail.png
---

<h1>Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method</h1>

<div class="blog-metadata">

>     <small>Published July 11, 2022.</small>

Reviewer 3: ```suggestion
    <small>Published Aug 2, 2022.</small>
```
